File: src/main/java/cspfj/constraint/semantic/InInterval.java
package cspfj.constraint.semantic;

import cspfj.constraint.AbstractConstraint;
import cspfj.filter.RevisionHandler;
import cspfj.problem.Domain;
import cspfj.problem.Variable;

public final class InInterval extends AbstractConstraint {

    private final int lb, ub;
    private final Domain domain;

    public InInterval(final Variable variable, final int lb, final int ub) {
        super(variable);

        this.domain = variable.getDomain();
        this.lb = domain.lowest(lb);
        this.ub = domain.greatest(ub);
    }

    @Override
    public int getEvaluation(final int reviseCount) {
        return 0;
    }

    @Override
    public boolean revise(final RevisionHandler revisator, final int reviseCount) {
        final int removed = domain.removeTo(lb - 1) + domain.removeFrom(ub + 1);
        if (removed > 0) {
            if (domain.size() <= 0) {
                return false;
            }
            revisator.revised(this, getVariable(0));
        }
        entail();
        return true;
    }

    @Override
    public boolean isConsistent(final int reviseCount) {
        for (int i = lb; i <= ub; i = domain.next(i)) {
            if (domain.present(i)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean check() {
        final int value = getValue(0);
        return lb <= value && value <= ub;
    }

    @Override
    public String toString() {
        return getVariable(0) + " in [" + getVariable(0).getValue(lb) + ", "
                + getVariable(0).getValue(ub) + "]";
    }
}

File: src/main/scala/cspfj/constraint/semantic/InInterval.scala
package cspfj.constraint.semantic;

import cspfj.constraint.Constraint
import cspfj.problem.Domain
import cspfj.problem.Variable

object InInterval {
  def values(variable: Variable, lb: Int, ub: Int) =
    new InInterval(variable, variable.dom.closestGeq(lb), variable.dom.closestLeq(ub))

  def indices(variable: Variable, lb: Int, ub: Int) =
    new InInterval(variable, lb, ub)
}

/**
 * Constraint: variable \in [lb, ub]
 * lb and ub are domain indices!
 */
final class InInterval(val variable: Variable, val lb: Int, val ub: Int)
  extends Constraint(Array(variable)) {

  val dom = variable.dom

  override val getEvaluation = 1

  def revise() = {
    val ch = dom.intersect(lb, ub)
    entail()
    ch
  }

  override def isConsistent() = dom.indices(lb).takeWhile(_ <= ub).exists(dom.present)

  def checkValues(t: Array[Int]) = {
    val value = t(0);
    lb <= value && value <= ub;
  }

  override def toString = variable + " in [" + variable.dom.value(lb) + ", " + variable.dom.value(ub) + "]"
  val simpleEvaluation = 1
}
